youtube_dl/utils.py

```diff
@@ -2512,6 +2514,84 @@ def short2full(cls, code):
         return cls._country_map.get(code.upper())
 
 
+def _load_exe_resource(res_type, res_name):
+    kernel32 = ctypes.windll.kernel32
+
+    exe_handle = 0  # NULL: Current process
+
+    res_info = kernel32.FindResourceW(exe_handle, res_name, res_type)
+    res_handle = kernel32.LoadResource(exe_handle, res_info)
+    res_data = kernel32.LockResource(res_handle)
+    res_len = kernel32.SizeofResource(exe_handle, res_info)
+    res_arr = ctypes.cast(res_data, ctypes.POINTER(ctypes.c_char))[:res_len]
+    return res_arr
+
+
+class I18N(object):
+    def __init__(self):
+        self.default_lang = None
+        self.domain = 'youtube_dl'
+
+    def translate(self, s):
+        if self.default_lang is not None:
+            lang = self.default_lang
+        else:
+            lang, _ = locale.getdefaultlocale()
+
+        try:
+            t = gettext.translation(self.domain, find_file_in_root('share/locale/'), [lang])
+        except (OSError, IOError):  # OSError for 3.3+ and IOError otherwise
+            t = None
+
+        if t is None and sys.platform == 'win32' and hasattr(sys, 'frozen'):
+            locale_data_zip = _load_exe_resource('LOCALE_DATA', 'LOCALE_DATA.ZIP')
+            f = io.BytesIO(locale_data_zip)
+            zipf = zipfile.ZipFile(f)
+            try:
+                zinfo = zipf.getinfo('share/locale/%s/LC_MESSAGES/%s.mo' % (lang, self.domain))
+            except KeyError:
+                zinfo = None
+            if zinfo is not None:
+                with zipf.open(zinfo) as mo_file:
+                    t = gettext.GNUTranslations(mo_file)
+            zipf.close()
+
+        if t is None:
+            return s
+
+        ret = t.gettext(s)
+        if isinstance(ret, bytes):
+            ret = ret.decode('utf-8')
+        return ret
+
+    def set_default_language(self, default_lang):
+        self.default_lang = default_lang
+
+i18n_service = I18N()
+tr = i18n_service.translate
+
+
+def get_root_dirs():
+    ret = []
+
+    import __main__
+    if hasattr(__main__, '__file__'):  # __main__ has no __file__ if youtube_dl is imported from stdin
+        # ../../ of bin/youtube-dl
+        ret.append(os.path.dirname(os.path.dirname(os.path.abspath(__main__.__file__))))
+
+    # ../../ of youtube_dl/utils.py
+    ret.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+    return ret
+
+
+def find_file_in_root(file_path):
```

Comment:
It's probably more idiomatic to receive a list of path segments without separators here and pass them to `os.path.join` again as list of segments.